mbed_build/_internal/config/config.py

```diff
@@ -2,47 +2,122 @@
 # Copyright (C) 2020 Arm Mbed. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 #
-"""Build configuration abstraction layer."""
-from typing import Any, Dict, List, Set, Optional
-from typing_extensions import TypedDict
+"""Build configuration representation."""
+import re
+import itertools
+from dataclasses import dataclass, field, fields
+from typing import Any, Dict, List, Optional, Set, Tuple
 
-from mbed_build._internal.config.config_layer import ConfigLayer
+from mbed_build._internal.config.source import Source
 
 
-class Config(TypedDict):
-    """Represents a build configuration."""
+@dataclass
+class Option:
+    """Representation of configuration option."""
 
-    settings: Dict[str, "Setting"]
-    target: "TargetOverrides"
+    value: Any
+    macro_name: Optional[str]
+    help_text: Optional[str]
+    set_by: str
 
+    @classmethod
+    def build(cls, data: Any, source: Source) -> "Option":
+        """Build configuration option from config entry value.
 
-class Setting(TypedDict):
-    """Represents a config setting."""
+        Config values are either complex data structures or simple values.
+        This function handles both.
+        """
+        if isinstance(data, dict):
+            return cls(
+                value=data.get("value"),
+                macro_name=data.get("macro_name"),
+                help_text=data.get("help"),
+                set_by=source.name,
+            )
+        else:
+            return cls(value=data, macro_name=None, help_text=None, set_by=source.name)
+
+    def set_value(self, value: Any, source: Source) -> None:
+        """Mutate self with new value."""
+        self.value = value
+        self.set_by = source.name
+        return self
```

Comment:
Looks like a typo here?

Reply:
? 🤔

Reply:
We shouldn't need to return `self` here, and the type hint says we return `None`.

Reply:
Ah, you're right. `self` is there for the sake of using this like a fluent interface. I'll fix the type hint.

Reply:
The bigger question is - how the heck does ci not fail?

Reply:
Answer: missing `__init__.py` file. https://github.com/ARMmbed/mbed-build/pull/27/commits/bf7c4bcc620e44ae69df62d46826d6d0cfc1d624